iris_grib/grib_phenom_translation/_gribcode.py

```diff
@@ -0,0 +1,191 @@
+# Copyright iris-grib contributors
+#
+# This file is part of iris-grib and is released under the BSD license.
+# See LICENSE in the root of the repository for full licensing details.
+'''
+Provide object to represent grib phenomena
+
+For use as cube attributes, freely convertible to+from strings
+'''
+from dataclasses import dataclass
+import re
+from typing import Optional
+
+
+def _invalid_edition(edition):
+    msg = (
+        f"Invalid grib edition, {edition!r}, for GRIBcode : "
+        "can only be 1 or 2."
+    )
+    raise ValueError(msg)
+
+
+def _invalid_nargs(args):
+    nargs = len(args)
+    msg = (
+        f"Cannot create GRIBCode from {nargs} arguments, "
+        f"GRIBCode({args!r}) : expects either 1 or 4 arguments."
+    )
+    raise ValueError(msg)
+
+
+_RE_PARSE_FOURNUMS = re.compile(4 * r'[^\d]*(\d*)')
+
+
+def _fournums_from_gribcode_string(edcn_string):
+    parsed_ok = False
+    nums_match = _RE_PARSE_FOURNUMS.match(edcn_string).groups()
+    if nums_match is not None:
+        try:
+            nums = [int(grp) for grp in nums_match]
+            parsed_ok = True
+        except ValueError:
+            pass
+
+    if not parsed_ok:
+        msg = ('Invalid argument for GRIBCode creation, '
+               '"GRIBCode({!r})" : '
+               'requires 4 numbers, separated by non-numerals.')
+        raise ValueError(msg.format(edcn_string))
+
+    return nums
+
+
+def GRIBCode(edition, *args, **kwargs):
+    """
+    Make an object representing a specific Grib phenomenon identity.
+
+    The class of the result, and the list of its properties, depend on whether
+    'edition' is 1 or 2.
+
+    One of :
+
+    * GRIBCode(edition=1, table_version, centre_number, number)
+    * GRIBCode(edition=2, discipline, category, number)
+
+    Either provides a string representation, and supports creation from:
+    keywords, another similar object; a tuple of numbers; or any string with 4
+    separate decimal numbers in it.
+    """
+    if edition is None:
+        _invalid_nargs(())
+
+    # Convert single argument to *args
+    if not args and not kwargs:
+        # Convert to a string and extract 4 integers.
+        # NOTE: this also allows input from a GRIBCode, or a plain tuple.
+        edition_string = str(edition)
+        edition, arg2, arg3, arg4 = \
+            _fournums_from_gribcode_string(edition_string)
+        args = [arg2, arg3, arg4]
+
+    # Check edition + select the relevant keywords for the edition
+    if edition not in (1, 2):
+        _invalid_edition(edition)
+
+    # Choose which actual type we will return.  This also determines the
+    # argument (keyword) names.
+    instance_cls = {1: GRIBCode1, 2: GRIBCode2}[edition]
+
+    # Convert all of (edition, *args) into **kwargs
+    if not args:
+        # Ignore that edition= is a required arg -- make it a kwarg
+        kwargs['edition'] = edition
```

Comment:
I don't think this block can be accessed?

Reply:
No it can, if the args are all keywords (except possibly edition).
E.G.`GRIBCode(2, discipline-0, category=5, number=7)`

Reply:
BTW I used pytest-cov to check I had covered everything
Except [this test line](https://github.com/SciTools/iris-grib/blob/22cea662f204b3056c788088156023938a98e52f/iris_grib/grib_phenom_translation/_gribcode.py#L38)
And I've now worked out why that never fails : see [other comment](https://github.com/SciTools/iris-grib/pull/402#discussion_r1567689111)